### ISSUE TITLE
Make auth plugins more pluggable

### ISSUE THREAD

**Issue by merlinthp** [thelounge/lounge#1981]
Issue #1953 mentions that in future auth plugins should be shipped separately from lounge itself.  That'll require a more formal auth plugin API, and some code being moved out of the main server code into plugins (mainly the local auth password change code).  Having taken a look at the current code, here's my idea on how to do this:

- Deprecate the public option, replace it with an auth option which specifies which auth plugin to use (a value of 'none' would equate to public).
- Add code to convert the current public and ldap.enable options to the appropriate auth setting, and write log entries encouraging the admin to update their config
- Don't have a hard-coded set of require() statements for auth plugins, just try to require() the plugin specified by the auth option
- Get rid of the isEnabled export in the existing auth modules, as they won't be needed
- Deprecate the ldap.enable option, and write log entries as with the public option
- Add a mechanism for auth modules to specify that they can change user passwords
- Change the ldapEnabled value sent to clients to "password change available" to make it auth module agnostic
- Move the local auth password change code out of the main server.js change-password handler into the local auth plugin

Thoughts?  I can work up a PR based on feedback.

**Comment by metsjeesus** [comment]
I have tried to make my own auth method over a year now, but allways fail on timings. Tried to make Promises, kind of works, but looks nice and has its own problems. Maybe to a event based system is better pluggable then list, names and promises?

**Comment by poVoq** [comment]
Yes this would be appreciated, especially if this would lead to other auth methods becoming available, such as an external user SQL database (for example from a forum) or OpenID Connect / Oauth2 support.

**Comment by xPaw** [comment]
If there was an update someone would have left a comment. If you need this, feel free to work on it.

**Comment by BrainWart** [comment]
I have interest in helping with this. I have not reviewed the entire auth flow yet, but my high level thoughts are to
1. Move auth plugin initialization to server.js
  - This will let us create new endpoints
2. Stack the authentication methods
  - If one fails fallback to the next? or maybe a selection for the auth method
3. Convert the existing auth plugins to "classes" and inject the app, config, etc through the initializer.

After a rough conversion to this, I plan to implement OIDC or import an external OIDC library and tie it in to see what might need to change for passwordless auth.

I am concerned how multiple providers might be handled by the internal user management so I'll be poking around there a bit to see how that might be handled securely.

**Comment by BrainWart** [comment]
I see the packages as a concept in install custom modifications now. I think I will still need to do a bit of the above, but if auth packages are the plan, I'll try to keep that in mind and see if I can allow auth to be inserted and removed dynamically.

**Comment by poVoq** [comment]
if you plan to implement Oauth2, please have a look at how Gamja does it, and make it compatible.

**Comment by brunnre8** [comment]
>please have a look at how Gamja does it,

Gamja has absolutely nothing to do with it, we're talking TL client --> TL server auth

**Comment by poVoq** [comment]
Yes, but there are IRC servers (Ergo) and bouncers (Soyu) that also support Oauth2 and it can be passed though to them and Gamja is Afaik currently the only implementation of that login flow. I agree though that TL needs to do it somewhat different as it has it's own server component, which Gamja doesn't.
